### PR TITLE
Specify the HTTP Gateway protocol

### PR DESCRIPTION
**Pull request by nomeata**
We have had the HTTP gateway for a while now, but 

* we didn’t have a public record of the expected Canister interface, with the authorative Candid fragment
* we already have multiple implementations (`icx-proxy` and the service worker), and people we looking at the implementations to understand how thing work.
* we see new features being implemented in one, but not the other implementation, and they start to diverge
* some features are only implementation-defined (e.g. the certification fallback to `index.html`)
* the interaction of new features (e.g. certificate, upgrade to update and streaming) is ill or at least vaguely defined.

So I think we urgently need a proper specification for the HTTP Gateway protocol, and a place to host this. The IC Interface Spec is a natural place, so here is a possible start.

This PR is descriptive, not prescriptive, with regard to the status quo: I (try to) specify the existing behavior of our Gateway implementations, not define a new and different protocol.

[Preview available](https://ic-spec-pr-preview.netlify.app/#http-gateway), pushed manually until #9 or similar has been fixed.

### CODE REVIEW

**Comment by nomeata** [comment]
Draft until external contributions are possible.

@paulyoung, can you also review this? It would be helpful if you could double check this. Who else should be pinged?

**Comment by nomeata** [comment]
@jplevyak, right now the spec says that all headers are passed through by the boundary node, but that’s not really true, is it? What header filtering/mangling/addition should we specify/document here?

File: spec/index.adoc
Comment (jzxchiang1) [inline spec/index.adoc:1970]: Typo: `futher` -> `further`

File: spec/index.adoc
Comment (jzxchiang1) [inline spec/index.adoc:1986]: Typo: `Canisters not using these features can completely`

`simplifiy` -> `simplify`

File: spec/index.adoc
Comment (jzxchiang1) [inline spec/index.adoc:2003]: Where is `personhood` specified?

I don't see it [here](https://github.com/dfinity/agent-js/blob/6e8c64cf07c7722aafbf52351eb0f19fcb954ff0/apps/sw-cert/src/sw/http_request.ts#L15).

Not really convinced this canister resolution belongs in the HTTP Gateway protocol, although it is good to know in general I suppose.

File: spec/index.adoc
Comment (jzxchiang1) [inline spec/index.adoc:2017]: Very important! The query parameters inclusion tripped me up.

File: spec/index.adoc
Comment (jzxchiang1) [inline spec/index.adoc:2025]: Probably clearer to say "with the same `HttpRequest`" instead of "with the same argument" IMO

File: spec/index.adoc
Comment (jzxchiang1) [inline spec/index.adoc:1959]: This paragraph confused me. I'm not exactly sure what the "system" is. Also, the "it" in "It is still included in this file" didn't make sense to me.

File: spec/index.adoc
Comment (jzxchiang1) [inline spec/index.adoc:2045]: Slightly off-topic, but I've always been confused exactly how this streaming actually gets implemented. Is it HTTP/1.1 chunked transfer encoding or HTTP/2 streaming?

Should the protocol described in this document talk about whether HTTP/1.1 or HTTP/2 (or both) are supported?

File: spec/index.adoc
Comment (jzxchiang1) [inline spec/index.adoc:2019]: I wonder if the protocol should specify which request headers are special and will be automatically processed by any Gateway implementation.

I suppose `Host` is an example of one, but how about `Cache-Control` or any others? icx-proxy doesn't do any caching AFAIK (although boundary nodes, which run icx-proxy, apparently do). Should these kind of features be specified in the protocol?

File: spec/http-gateway.did
Comment (jzxchiang1) [inline spec/http-gateway.did:14]: Typo: `upgrade` -> `update`

File: spec/http-gateway.did
Comment (jzxchiang1) [inline spec/http-gateway.did:25]: Unfortunately, there's a Candid bug which in practice will cause a runtime error if you define the callback like this.

See this [comment](https://forum.dfinity.org/t/cryptic-error-from-icx-proxy/6944/9?u=jzxchiang) by @chenyan-dfinity for details.

Perhaps the protocol can remain "pristine" but if users want code that actually works, they will need to change the type to:

`callback: func (?T) -> (opt StreamingCallbackHttpResponse);`

File: spec/index.adoc
Comment (jzxchiang1) [inline spec/index.adoc:2047]: I don't feel qualified to review the certification part at the moment, so I'll defer this for someone else.

File: spec/http-gateway.did
Comment (jzxchiang1) [inline spec/http-gateway.did:33]: I think it'd be clearer to make reference to the streaming callback here, but I understand why you didn't. (It's optional and also dynamically specified in `HttpResponse`.)

**Review by jzxchiang1 — CHANGES_REQUESTED** [review]

**Comment by jplevyak** [comment]
@frederikrothenberger and @Daniel-Bloom-dfinity are current working this.

File: spec/index.adoc
Comment (nomeata) [reply spec/index.adoc:2003]: I saw it for the first time yesterday in https://github.com/dfinity/ic/blob/a254116be472f81b85b76429ac8d987f80dd4411/typescript/service-worker/src/sw/http_request.ts#L17. I'll leave it to the Dfinity people to say if this should be there or not. But it demonstrates the need for a authorative list, at least until we have a dynamic system in place.

File: spec/index.adoc
Comment (nomeata) [reply spec/index.adoc:2019]: Yes, definitely! Unfortunately I don't know the precise state of affairs here, and hope that @jplevyak can fill in the blanks.

Although maybe some things need not be documented. If the boundary nodes acts on certain headers in a HTTP spec compliant way, e.g. while caching, then that should be fine even if it isn't written here. just as if some user uses their ISP's proxy and that caches.

File: spec/http-gateway.did
Comment (nomeata) [reply spec/http-gateway.did:25]: Is that just due to bugs in some implementation that will be fixed in due time? Then I'd rather keep it as designed here, and leave temporary work-arounds out.

File: spec/http-gateway.did
Comment (nomeata) [reply spec/http-gateway.did:33]: The name is arbitrary, right? So I wouldn't quite know how to include it.

File: spec/index.adoc
Comment (nomeata) [reply spec/index.adoc:2045]: This streaming is _only_ concerned with getting the data from the canister to the Gateway, to overcome the message size limit, so it's independent of any HTTP-level streaming. Note that with certification (as it is specified now), the Gateway can't even stream out the data until it has received it all. In the uncertified case it could stream it out (on the TCP level, or the HTTP level) - I don't care and leave that as as implementation choice. Note that if the Gateway is the browser itself it also doesn't make sense to stream.

But it's a good point and I'll clarify!

File: spec/index.adoc
Comment (nomeata) [reply spec/index.adoc:1959]: I rewrote it. Better? Or is this not a helpful paragraph?

File: spec/http-gateway.did
Comment (nomeata) [reply spec/http-gateway.did:14]: Unfortunately(?) this field _is_ called `upgrade`, as in “upgrade the call from query to an update call”. Very confusing… oh well.

See <https://github.com/dfinity/icx-proxy/pull/6/files#diff-42cb6807ad74b3e201c5a7ca98b911c5fa08380e942be6e4ac5807f8377f87fcR259>

File: spec/index.adoc
Comment (nomeata) [inline spec/index.adoc:2035]: Is the boundary node mangling or filtering the headers in any way that should be noted here?

**Comment by origyndev** [comment]
If we are making suggestions here I’d love for there to be a “return and upgrade” feature that returns the result of the query to the client but then calls update for recording the serving of the request in the canister. The reply to this call can be discarded as the request has already been served.

**Comment by nomeata** [comment]
> If we are making suggestions here

I suggest we don't, for now. Let's focus in finalizing a spec for the status quo. Might be hard enough.

Once thats done additional changes can be nicely discussed in their own PRs against master. Or maybe on the forum first.

File: spec/http-gateway.did
Comment (chenyan-dfinity) [reply spec/http-gateway.did:25]: It's only an implementation problem. There is a more recent workaround here by abusing serde's `ignored_any`: https://github.com/dfinity/agent-rs/blob/main/ic-utils/src/interfaces/http_request.rs#L231

I think the real problem here is that `T` is undefined, it's not a proper Candid file.

Comment (nomeata) [reply spec/http-gateway.did:25]: Yeah, but that's too late to avoid now, right?

Comment (chenyan-dfinity) [reply spec/http-gateway.did:25]: I guess the solution is to catch up on the Candid side to add generics? In the Rust library, we kind of overcome this by deserializing the value into an untyped `IDLValue`.

In the meantime, we probably need to add a comment here to not confuse the reader.

Comment (nomeata) [reply spec/http-gateway.did:25]: I wrote this in the adoc:

>  This type signature is a type signature _scheme_, and the canister can choose any type for the token type `T` used for streaming. The Gateway handles that type generically.

But yes, maybe better moved into the Candid file proper?

File: spec/index.adoc
Comment (frederikrothenberger) [reply spec/index.adoc:2019]: Imho, the gateway should interfere with the HTTP interaction as little as possible. There is one issue though, which is that the service worker is unable to pass on `Set-Cookie` headers, as they are `forbidden`: https://fetch.spec.whatwg.org/#forbidden-response-header-name

Comment (nomeata) [reply spec/index.adoc:2019]: This only affects service workers, but not Gateways implemented as proxies or IC-aware browsers, right?

In that case I would not forbid them here, but maybe leave a NOTE that not all Gateway implementations may be able to pass these headers on?

File: spec/index.adoc
Comment (Daniel-Bloom-dfinity) [reply spec/index.adoc:2035]: The filtering we have is in [dfinity/ic/ic-os/boundary-guestos/rootfs/etc/nginx/conf.d/001-ic-nginx.conf](https://github.com/dfinity/ic/blob/master/ic-os/boundary-guestos/rootfs/etc/nginx/conf.d/001-ic-nginx.conf#L154).

The short of it is we add these response headers (and thus their implications) to every HTTP call response:

```access-control-allow-credentials: true
access-control-allow-origin: *
access-control-allow-methods: GET, POST, HEAD, OPTIONS
access-control-allow-headers: DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Cookie
access-control-expose-headers: Content-Length,Content-Range
x-cache-status: MISS
```

File: spec/http-gateway.did
Comment (nomeata) [reply spec/http-gateway.did:25]: Hmm, `reserved` isn’t quite right (that is a specific Candid type) But maybe
```
type Token = /* insert service-specific type here */
```
or something like that

File: spec/index.adoc
Comment (nomeata) [reply spec/index.adoc]: Done. Feel free to suggest edits to the wording if it should be further refined.

File: spec/http-gateway.did
Comment (nomeata) [reply spec/http-gateway.did]: Refined, and added a stern warning

File: spec/index.adoc
Comment (nomeata) [reply spec/index.adoc:2035]: I guess the question is: Why set these? Shouldn’t the canister have control over them?

**Comment by nomeata** [comment]
I’ve manually pushed a [rendering of this to `netlify`](https://ic-spec-pr-preview.netlify.app/#http-gateway), to get a preview until #9 or similar has been fixed.

File: spec/http-gateway.did
Comment (crusso) [reply spec/http-gateway.did:14]: Probably to late to change, but would `promote` be a better name?

**Comment by paulyoung** [comment]
I'll try to review this soon but in the meantime I wanted to share https://forum.dfinity.org/t/boundary-node-http-response-headers/10747 in case its of interest or provides some additional insight.

> Who else should be pinged?

Based on the above I think @lastmjs would be interested in this.

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:1977]:
```suggestion
The following interface description, in https://github.com/dfinity/candid/blob/master/spec/Candid.md[Candid syntax], describes the expected Canister interface. You can also link:{attachmentsdir}/http-gateway.did[download the file].
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:1984]:
```suggestion
NOTE: Canisters not using these features can completely leave out the `streaming_strategy` and/or `upgrade` fields in the `HttpResponse` they return, due to how Candid subtyping works. This might simplify their code.
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2004]:
```suggestion
3. Else, if `<name>` is a valid textual encoding of a principal, use that principal as the canister id.
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2014]:
```suggestion
* The `method` field contains the HTTP method (e.g. `HTTP`), in upper case.
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2036]:
```suggestion
NOTE: Not all Gateway implementations may be able to pass on all forms of headers. In particular, Service Workers are unable to pass on the `Set-Cookie` header.
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2040]:
```suggestion
HTTP Gateways may add additional headers. In particular, the following headers may be set:
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2056]:
```suggestion
The HTTP Gateway protocol has provisions to transfer further chunks of the body data from the canister to the HTTP Gateway, to overcome the message limit of the Internet Computer. This streaming protocol is independent of any possible streaming of data between the HTTP Gateway and the HTTP client. The gateway may assemble the response in whole before passing it on, or pass the chunks on directly, on the TCP or HTTP level, as it sees fit. When the Gateway is <<http-gateway-certification,certifying the response>>, it must not pass on uncertified chunks.
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2062]:
```suggestion
2. Else, it makes a query call to the given method, passing the `token` value given in the `streaming_strategy` as the argument.
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2066]:
```suggestion
WARNING: The type of the `token` value is chosen by the canister; the HTTP Gateway obtains the Candid type of the encoded message from the canister, and uses it when passing the token back to the canister. This generic use of Candid is not covered by the Candid specification, and may not be possible in some cases (e.g. when using “future types”). Canister authors may have to use “simple” types.
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2012]:
```suggestion
The HTTP request is encoded into the `HttpRequest` Candid structure.
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2076]:
```suggestion
2. The value of the header must be a structured header according to RFC 8941 with fields `certificate` and `tree`, both being byte sequences.
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2078]:
```suggestion
3. The `certificate` must be a valid certificate as per <<certification>>, signed by the root key. If the certificate contains a subnet delegation, the delegation must be valid for the given canister. The timestamp in `/time` must be recent. The subnet state tree in the certificate must reveal the canister’s <<state-tree-certified-data,certified data>>.
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2080]:
```suggestion
4. The `tree` must be a hash tree as per <<certification-encoding>>.
```

File: spec/index.adoc
Comment (paulyoung) [inline spec/index.adoc:2086]:
```suggestion
7. That leaf must contain the SHA-256 hash of the _decoded_ body.
```

File: spec/index.adoc
Comment (Daniel-Bloom-dfinity) [reply spec/index.adoc:2034]: The short answer is "it's complicated". The longer answer is "because headers are not yet certified". As such we currently limit headers a bit to improve security. Once we get certified headers, we should not have any more need for this limitation.

File: spec/http-gateway.did
Comment (nomeata) [reply spec/http-gateway.did:14]: Probably. Or `resend_as_update_call`. Or, in line with the pattern for the streaming callback, it could even pass a `func` ref.

In a way it's not too late: we can change this while keeping backward conpat in the implementations (there is only one so far) relatively easily (the proxy can simply keep both in it's internal expected type). I just wonder if we should do it with this PR, or maybe as a separate step, to get this in first.

File: spec/index.adoc
Comment (nomeata) [reply spec/index.adoc:2034]: Ok, good enough for this PR I guess.

**Comment by paulyoung** [comment]
Pinging @SuddenlyHazel for visibility since this set of functionality came up in conversations today.

**Comment by nomeata** [comment]
Ok, with the CLA requirement out of the way, this _could_ now be merged. Who on the DFINITY side is owning the process of shepherding and eventually merging community contributions? @Dfinity-Bjoern?

**Comment by jwiegley** [comment]
@nomeata It is Björn and myself.

**Comment by ali-dfinity** [comment]
update cla

File: spec/index.adoc
Comment (frederikrothenberger) [inline spec/index.adoc:2078]: Currently neither icx-proxy nor the service worker check the certificate `/time` (which is bad). Can you give some more specific measure of "recent"? 5 minutes?

Comment (nomeata) [reply spec/index.adoc:2078]: That’s probably for the security team to decide. 5 min should be ample (until we get into edge node caching, then things become tricky, because a cache is hardly different from an attacker serving old data).

**Review by frederikrothenberger — APPROVED** [review]
Thanks for writing this spec.

**Review by Dfinity-Bjoern — APPROVED** [review]
That's great! Thank you very much @nomeata for contributing this!

**Comment by nomeata** [comment]
Note that I am not allowed to merge, so someone else will have to press that button (and ideally presses the “Squash and merge” button and manually cleans up the PR description or copies the PR description into the commit message – mergify would do that for us but that’s not enabled at the moment.)

**Review by Daniel-Bloom-dfinity — APPROVED** [review]

**Pull request merged**
